### PR TITLE
Extract additional types

### PR DESCRIPTION
**Pull request by willbakst**
- Added support for extracted `BaseTool` variants
- Added support for extracting base types (e.g. str, int, etc.) with `Prompt.extract(str)`

### CODE REVIEW

File: mirascope/base/types.py [hunk 1]
```diff
@@ -1,11 +1,46 @@
...
 
 from .tools import BaseTool
 
+BaseType = Union[
+    str,
+    int,
+    float,
+    bool,
+    list,
+    dict,
```
Comment (brianpark0126) [inline mirascope/base/types.py:28]: extracting a dict gives Pydantic validation error:
pydantic_core._pydantic_core.ValidationError: 1 validation error for DictTool
value
  Field required [type=missing, input_value={'tool_call': ChatComplet...ool'), type='function')}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/missing

File: mirascope/openai/prompt.py [hunk 2]
```diff
@@ -183,7 +187,19 @@ async def async_stream(self) -> AsyncGenerator[OpenAIChatCompletionChunk, None]:
                 chunk=chunk, tool_types=tools if tools else None
             )
 
+    @overload
+    def extract(self, schema: Type[BaseTypeT], retries: int = 0) -> BaseTypeT:
```
[the remark this answers is not in the document]
Comment (willbakst) [reply mirascope/openai/prompt.py:191]: these are just overloads for type hints. The main docstring for the non-overloaded definition shows up for all overloaded definitions :)

File: mirascope/base/types.py (hunk 1 above)
Comment (willbakst) [reply mirascope/base/types.py]: Ok after further testing and research, it appears that it just doesn't work with dict, so I've removed support for dict as a base class (the tool just isn't populated in this case).

Will punt to further research.

File: mirascope/base/types.py [hunk 3]
```diff
@@ -1,11 +1,46 @@
...
 from .tools import BaseTool
 
+BaseType = Union[
+    str,
+    int,
+    float,
+    bool,
+    list,
+    set,
+    tuple,
```
Comment (brenkao) [inline mirascope/base/types.py:29]: what about dict? or bytes?

Comment (willbakst) [reply mirascope/base/types.py:29]: not handled. dict just doesn't work, bytes is also hard and not necessarily that useful, so I've excluded.

We can look to add them later, but it wasn't a simple addition

**Review by brenkao — APPROVED** [review]

**Pull request merged**